### PR TITLE
[core]: introduce a common openstack client package

### LINKED ISSUE #993: Move out OpenStack new client logic to a separate package
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**The binaries affected**:

- [x] openstack-cloud-controller-manager
- [ ] cinder-csi-plugin
- [ ] k8s-keystone-auth
- [ ] client-keystone-auth
- [ ] octavia-ingress-controller
- [ ] manila-csi-plugin
- [ ] manila-provisioner
- [ ] magnum-auto-healer
- [ ] barbican-kms-plugin

**What happened**:

Currently openstack new client logic is defined in openstack-cloud-controller-manager, but used by all openstack related controllers.

**What you expected to happen**:

A better place for this logic is a separate package.

### PR DESCRIPTION
**Pull request by kayrus**
**What this PR does / why we need it**:

This PR optimizes the package structure. As an outcome some binary sizes are much smaller.

**Which issue this PR fixes(if applicable)**:
fixes #993

**Special notes for reviewers**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by kayrus** [comment]
@lingxiankong @jichenjc ready for review

**Comment by kayrus** [comment]
@jichenjc any way to merge this without a test? Last commit just fixed a simple `go vet` warning, but now 4 tests are failing.

**Comment by jichenjc** [comment]
we can overwrite the test result, but I think it's a refactory PR so not a urgent that we have to merge now?
can we wait for openlab issue fix then we can retest this one? @kayrus

**Comment by kayrus** [comment]
@jichenjc it is an important refactoring before I introduce the support for retry logic. See https://github.com/gophercloud/gophercloud/pull/2097 for details.

**Comment by jichenjc** [comment]
@kayrus ok, as you can see openlab is taking actions so maybe we can wait for 1 day from them? 
I tried 2 hours ago still failing, maybe I will ask them shorting again

**Comment by jichenjc** [comment]
@kayrus do we need bump gophercloud version?

**Comment by kayrus** [comment]
@jichenjc not yet. I'll bump it in a follow-up PR. Besides there could be conflicts with other PRs, which bump gophercloud, e.g. #1365

**Comment by kayrus** [comment]
cloud-provider-openstack-acceptance-test-e2e-conformance-stable-branch-v1.19

**Comment by kayrus** [comment]
@ramineni @lingxiankong ready for review

**Review by lingxiankong — CHANGES_REQUESTED** [review]
Overalll looks good, thanks for the work! One missing part is the `.zuul.yaml` file

File: pkg/autohealing/cloudprovider/register/register.go [hunk 1]
```diff
@@ -27,38 +27,37 @@ import (
...
 func registerOpenStack(cfg config.Config, kubeClient kubernetes.Interface) (cloudprovider.CloudProvider, error) {
-	client, err := openstack_provider.NewOpenStackClient(&cfg.OpenStack, "magnum-auto-healer")
+	client, err := client.NewOpenStackClient(&cfg.OpenStack, "magnum-auto-healer")
 	if err != nil {
 		return nil, err
 	}
 
+	eoOpts := gophercloud.EndpointOpts{
+		Region:       cfg.OpenStack.Region,
+		Availability: cfg.OpenStack.EndpointType,
```
[the remark this answers is not in the document]
Comment (kayrus) [reply pkg/autohealing/cloudprovider/register/register.go:41]: Added an entry in docs.

**Comment by lingxiankong** [comment]
> One missing part is the `.zuul.yaml` file

I didn't see this is resolved.

**Comment by kayrus** [comment]
> > One missing part is the .zuul.yaml file

> I didn't see this is resolved.

@lingxiankong what do you mean? I haven't touched any zuul.yaml files and they're not in this PR.

**Comment by lingxiankong** [comment]
The only change related to k8s-keystone-auth is a package import in the file `pkg/identity/keystone/token_getter.go`, so it's safe to override the keystone-authentication-authorization job.

/override openlab/cloud-provider-openstack-acceptance-test-keystone-authentication-authorization

**Comment by lingxiankong** [comment]
All my comments have been resolved.

/lgtm

**Comment by kayrus** [comment]
@lingxiankong  re the `cloud-provider-openstack-acceptance-test-keystone-authentication-authorization` job. do you have any idea why it fails now, but didn't fail earlier?

**Comment by kayrus** [comment]
@ramineni could you please check this PR?

**Comment by kayrus** [comment]
rebased due to conflicts in `pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go`

**Comment by lingxiankong** [comment]
> @lingxiankong re the `cloud-provider-openstack-acceptance-test-keystone-authentication-authorization` job. do you have any idea why it fails now, but didn't fail earlier?

No idea, I need to add a sleep in the script and debug when I have some time.

**Comment by kayrus** [comment]
@lingxiankong @ramineni all tests have passed

File: pkg/client/client.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
+/*
```
Comment (ramineni) [inline pkg/client/client.go:1]: @kayrus we can add under pkg/openstack/client.go ?

File: pkg/client/service_client.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (ramineni) [inline pkg/client/service_client.go:1]: ditto

**Review by ramineni — COMMENTED** [review]
@kayrus Thanks for working on this. Looks good to me mostly.
Some suggestions inline. And also, I suppose we can remove the prefix openstack from the files under that dir.

File: pkg/client/client.go (hunk 2 above)
Comment (kayrus) [reply pkg/client/client.go:1]: No, the whole PR is to split openstack and client packages. If I add it back, the binary sizes will increase again.

**Comment by kayrus** [comment]
> Some suggestions inline. And also, I suppose we can remove the prefix openstack from the files under that dir.

updated filenames

File: pkg/client/client.go (hunk 2 above)
Comment (ramineni) [reply pkg/client/client.go:1]: Ah , got it. Its common client package, I assumed common openstack package  . 
But our ultimate goal is to have common openstack package as well , as there is much redundant code there too. Are you planning to work on that as well?

Comment (kayrus) [reply pkg/client/client.go:1]: > common openstack package

hm, no. current openstack package is primarily used in occm. other packages don't depend on it.

Comment (kayrus) [reply pkg/client/client.go:1]: My first step is to work on supporting the `Retry-After` throttling header.

**Comment by kayrus** [comment]
@ramineni @lingxiankong any other comments?

**Comment by lingxiankong** [comment]
/lgtm

I'm ok with this change. Let's merge it then other PRs could be rebased and reviewed.

**Comment by ramineni** [comment]
Thanks.
/lgtm
/approve

**Pull request merged**